### PR TITLE
Mxyng/decode

### PR DESCRIPTION
**Pull request by mxyng**

### CODE REVIEW

File: llm/gguf.go [hunk 1]
```diff
@@ -87,38 +86,41 @@ func (llm *ggufModel) NumKV() uint64 {
...
-func (llm *ggufModel) ModelType() ModelType {
+func (llm *ggufModel) ModelType() string {
 	switch llm.ModelFamily() {
-	case ModelFamilyLlama:
-		blocks, ok := llm.kv["llama.block_count"].(uint32)
-		if ok {
-			return ModelType(blocks)
+	case "llama":
+		if blocks, ok := llm.kv["llama.block_count"].(uint32); ok {
+			if numGQA, ok := llm.kv["llama.num_gqa"].(uint32); ok && numGQA == 0 {
```
Comment (BruceMacD) [inline llm/gguf.go:102]: what about grouped query attention being 0 indicates that the model is 70B?

**Review by BruceMacD — APPROVED** [review]
the logic and clean-up looks good, my main confusion is around how the gqa indicates the parameter size.

**Review by BruceMacD — APPROVED** [review]
the logic and clean-up looks good, my main confusion is around how the gqa indicates the parameter size.

**Review by BruceMacD — APPROVED** [review]
the logic and clean-up looks good, my main confusion is around how the gqa indicates the parameter size.

File: llm/gguf.go [hunk 2]
```diff
@@ -87,38 +86,41 @@ func (llm *ggufModel) NumKV() uint64 {
...
-func (llm *ggufModel) ModelType() ModelType {
+func (llm *ggufModel) ModelType() string {
 	switch llm.ModelFamily() {
-	case ModelFamilyLlama:
-		blocks, ok := llm.kv["llama.block_count"].(uint32)
-		if ok {
-			return ModelType(blocks)
+	case "llama":
+		if blocks, ok := llm.kv["llama.block_count"].(uint32); ok {
+			if numGQA, ok := llm.kv["llama.num_gqa"].(uint32); ok && numGQA == 8 {
```
Comment (mxyng) [inline llm/gguf.go:102]: Need to confirm this kv exists. If it doesn't, calculate it from other kvs

File: server/images.go [hunk 3]
```diff
@@ -815,14 +821,14 @@ func formatParams(params map[string][]string) (map[string]interface{}, error) {
 						return nil, fmt.Errorf("invalid float value %s", vals)
 					}
 
-					out[key] = floatVal
+					out[key] = float32(floatVal)
```
Comment (mxyng) [inline server/images.go:824]: `ParseFloat` returns a float64. Since this is explicitly handling float32s, type convert it

File: server/images.go [hunk 4]
```diff
@@ -815,14 +821,14 @@ func formatParams(params map[string][]string) (map[string]interface{}, error) {
...
 				case reflect.Int:
-					intVal, err := strconv.ParseInt(vals[0], 10, 0)
+					intVal, err := strconv.ParseInt(vals[0], 10, 64)
 					if err != nil {
 						return nil, fmt.Errorf("invalid int value %s", vals)
 					}
 
-					out[key] = intVal
+					out[key] = int(intVal)
```
Comment (mxyng) [inline server/images.go:831]: Same for int

File: llm/gguf.go (hunk 1 above)
Comment (mxyng) [reply llm/gguf.go]: That was a typo. It should be 8. The reason is 65B of llama and 70B of llama2 both have 80 layers. However 70B llama2 has grouped query attention so we can use that differentiate between the two

**Pull request merged**